pycontrails/datalib/spire.py

```diff
@@ -737,3 +754,663 @@ def _downsample_flight(
+
+        departure_lon = row["departure_airport_lon"]
+        departure_lat = row["departure_airport_lat"]
+        departure_alt_ft = row["departure_airport_alt_ft"]
+        if any(
+            [
+                np.isnan(departure_lon),
+                np.isnan(departure_lat),
+                np.isnan(departure_alt_ft),
+            ]
+        ):
+            return np.nan
+
+        return cls._calc_distance_m(
+            lon_0=row["longitude"],
+            lat_0=row["latitude"],
+            alt_ft_0=row["altitude_baro"],
+            lon_f=departure_lon,
+            lat_f=departure_lat,
+            alt_ft_f=departure_alt_ft,
+        )
+
+    @classmethod
+    def _calc_dist_to_arrival_airport(cls, row: pd.Series) -> float:
+        """
+        Calculate the distance from a given waypoint to the arrival airport.
+
+        Returns
+        -------
+        distance in meters to arrival airport.
+        np.nan if it cannot be calculated.
+        """
+
+        if "arrival_airport_lat" not in row.index:
+            raise ValueError("field arrival_airport_lat must be present.")
+        if "arrival_airport_lon" not in row.index:
+            raise ValueError("field arrival_airport_lon must be present.")
+        if "arrival_airport_alt_ft" not in row.index:
+            raise ValueError("field arrival_airport_alt_ft must be present.")
+
+        arrival_lon = row["arrival_airport_lon"]
+        arrival_lat = row["arrival_airport_lat"]
+        arrival_alt_ft = row["arrival_airport_alt_ft"]
+        if any([np.isnan(arrival_lon), np.isnan(arrival_lat), np.isnan(arrival_alt_ft)]):
+            return np.nan
+
+        return cls._calc_distance_m(
+            lon_0=row["longitude"],
+            lat_0=row["latitude"],
+            alt_ft_0=row["altitude_baro"],
+            lon_f=arrival_lon,
+            lat_f=arrival_lat,
+            alt_ft_f=arrival_alt_ft,
+        )
+
+    def _calculate_additional_fields(self):
+        """
+        Add additional columns to the provided dataframe.
+
+        These additional fields are needed to apply the validation ruleset.
+        """
+        self._df = self._df.assign(
+            elapsed_seconds=[
+                self._rolling_time_delta_seconds(window) for window in self._df.rolling(window=2)
+            ],
+        )
+        self._df = self._df.assign(
+            elapsed_distance_m=[
+                self._rolling_distance_meters(window) for window in self._df.rolling(window=2)
+            ],
+        )
+        self._df = self._df.assign(
+            ground_speed_m_s=self._df["elapsed_distance_m"]
+            .divide(self._df["elapsed_seconds"])
+            .replace(np.inf, np.nan)
+        )
+        self._df = self._df.assign(
+            rocd_fps=[self._rolling_rocd_fps(window) for window in self._df.rolling(window=2)]
+        )
+
+        if len(self._df["arrival_airport_icao"].value_counts()) > 1:
```

Comment:
`len(self._df["arrival_airport_icao"].value_counts())` -> `self._df["arrival_airport_icao"].nunique()`

(here and below)